### ISSUE TITLE
Add support for SNI with TLS

### ISSUE THREAD

**Issue by a-entin** [cockroachdb/cockroach#79163]
Postgres 14 added support for SNI.
CockroachDB strives for feature parity with [Postgres 13](https://www.cockroachlabs.com/docs/stable/postgresql-compatibility.html).

Support for SNI with TLS makes Istio configuration much easier.

cc: @knz, @JeffSwenson 

Jira issue: CRDB-14613

**Comment by knz** [comment]
thanks alex.

Postgres' SNI Support is documented here:

https://www.postgresql.org/docs/current/libpq-connect.html
![image](https://user-images.githubusercontent.com/642886/161323085-d90d0d03-f677-4115-84fa-265e53ac5332.png)

This looks like only a client-side feature?

Here's the patch that added said support:  [ssl-sni-v1a.patch](https://www.postgresql.org/message-id/attachment/95468/ssl-sni-v1a.patch)
In a nutshell:

- the above doc
- client-side, the above documented flag sets the SNI header in the TLS handshake to be equal to the hostname configured with `host` (client-side)

... and nothing else!

So, apparently... **this is not a Postgres feature, it's merely a libpq feature** (where libpq is the C client driver).

**Comment by knz** [comment]
All this suggests we don't require any additional work in CockroachDB to announce "OK we support this".

@JeffSwenson what do you think? Would it be OK to have a checkmark in docs saying "CRDB and CC already support this"?

**Comment by JeffSwenson** [comment]
I see a two distinct use cases for SNI:
1. Put an SNI aware proxy like istio in front of CRDB. Route to CRDB using the SNI information. This is something we already support. When I look at CC production logs I see a lot of `unknown configuration parameter: ‹"sni-server"›`. We may want to hide that error message, but it suggests setting SNI is compatible with CRDB.
2. Put an SNI aware proxy like istio between CRDB instances. SNI is used to route intra-cluster traffic. I do not know if this works out of the box. GRPC probably does the right thing automatically.

**Comment by bdarnell** [comment]
The underlying request that led to the filing of this issue is about supporting istio. Unfortunately istio's documentation confuses people into thinking that SNI is the issue, when in fact the issue is whether the protocol is TLS-first or not. HTTPS is TLS-first (i.e. the TLS ClientHello is the first data sent on the connection), but pgwire is not. So whether SNI is used or not, it's not going to help istio. Istio documents this as ["server-first protocols"](https://istio.io/latest/docs/ops/deployment/requirements/#server-first-protocols), but this is also confusing because pgwire is client-first but it's still not TLS-first.

So if we want to make this issue about supporting istio, it would mean adopting a TLS-first protocol. Since pgwire v3 is not TLS-first, this would mean either working with the postgres development community to introduce a v4 (or at least v3.1) postgres protocol which could be TLS-first, or to move away from compatibility with existing postgres drivers and introduce cockroach-specific drivers which could use our own TLS-first protocol. Needless to say either path would be a large undertaking. 

If we want to make this issue about "supporting SNI", what exactly would that mean? Right now clients can pass SNI and we ignore it (and this has always been true), which doesn't seem like something that's worth documenting as "supported". In the multi-tenant SQL proxy we have a good use case for SNI to handle routing without options or database name hacks, so that's what "supporting SNI" means to me. And in this case it would be something to document since the connection strings for drivers that support SNI would be different from those that don't.

**Comment by yvmori1** [comment]
Hi!
I'm the client who brought this request to @a-entin in the first place. Our goal is to use istio with wildcard certs and DNS resolution while keeping the CRDB port to 26257 for all instances.

Example:
instance1.myenv.domain.com:26257 --> ISTIO (virtual Service: SNI host == instance1.myenv.domain.com then route to serivce-crdb1.mynamespace.myenv.cluster.local)
 ```yaml
---
apiVersion: networking.istio.io/v1beta1
kind: VirtualService
metadata:
  name: instance0-console
  namespace: myenv
spec:
  hosts:
    - instance0.myenv.k8s.domain.com
  gateways:
  - istio-system/coackroach-gw
  tls:
  - match:
    - uri:
        prefix: /
      sniHosts:
      - instance0.myenv.k8s.domain.com
    route:
    - destination:
        host: instance0-public.cockroachdb.myenv.svc.cluster.local
        port:
          number: 26257
```

**NOTE:**
After reading this thread, the istio documentation (again) and the PostgreSQL Doc pertaining to SNI. Now, I understand that ISTIO would have to understand pgwrie protocol not only support TLS SNI. I might have been quick to fire the question.

Still it would be so nice to get that feature working. ;-)

**Comment by knz** [comment]
I'm going to close this issue, based on the following reasoning (taking from Ben's earlier answer):

> if we want to make this issue about supporting istio, it would mean adopting a TLS-first protocol. Since pgwire v3 is not TLS-first, this would mean either working with the postgres development community to introduce a v4 (or at least v3.1) postgres protocol which could be TLS-first, or 

With regards to PostgreSQL compatibility, the CockroachDB project has historically been a follower to pg innovation, not a driver. We currently do not have plans to change this (especially in light of the point below). So we are not going to proactively approach the pg community to make this happen.

@yvmori1 I would invite you to approach the pg community and talk about this.

If/when the pg protocol evolves to become TLS-first, we will naturally extend CockroachDB to follow pg's lead. We do not need an issue to remain open in GH for this to happen.

> to move away from compatibility with existing postgres drivers and introduce cockroach-specific drivers which could use our own TLS-first protocol. 

This, in fact, is a likely future for the CockroachDB project. And here, we do not need an issue open for SNI compatibility either. If/when CockroachDB grows its own driver ecosystem, it is a near certainty the protocol will be TLS-first and support SNI.

**Comment by knz** [comment]
@jtsiros FYI
